knowledge-base/scheduler-kb-context-menu-on-appointments-and-empty-slots.md

```diff
@@ -0,0 +1,367 @@
+    private DateTime StartDate { get; set; } = DateTime.Today;
+    private SchedulerView CurrView { get; set; } = SchedulerView.Week;
+    private DateTime DayStart { get; set; } = DateTime.Today;
+
+    private SchedulerAppointment LastClickedAppointment { get; set; }
+    private SchedulerSlotTemplateContext LastClickedEmptySlot { get; set; }
+    private TelerikContextMenu<ContextMenuItem> TheItemContextMenu { get; set; }
+    private TelerikContextMenu<ContextMenuItem> TheSlotContextMenu { get; set; }
+
+    private List<ContextMenuItem> MenuItems = new List<ContextMenuItem>()
+        {
+            new ContextMenuItem
+            {
+                Text = "Delete",
+                CommandName = "delete",
+                Icon = SvgIcon.X
+            },
+            new ContextMenuItem
+            {
+                Text = "Toggle Important",
+                CommandName = "toggleimportant",
+                Icon = SvgIcon.Pencil
+            }
+        };
+    private List<ContextMenuItem> MenuSlots = new List<ContextMenuItem>()
+        {
+            new ContextMenuItem
+            {
+                Text = "Create appointment",
+                CommandName = "create",
+                Icon = SvgIcon.File
+            }
+        };
+
+    //handle ContextMenu for Item
+    private async Task ShowItemContextMenu(MouseEventArgs e, SchedulerAppointment appt)
+    {
+        LastClickedAppointment = appt;
+        PrepareMenuItems(LastClickedAppointment);
+        await TheItemContextMenu.ShowAsync(e.ClientX, e.ClientY);
+    }
+
+    private void PrepareMenuItems(SchedulerAppointment appt)
+    {
+        MenuItems[0].Disabled = appt.IsImportant;
+    }
+
+    private async Task MenuClickItemHandler(ContextMenuItem clickedItem)
+    {
+        if (!string.IsNullOrEmpty(clickedItem.CommandName) && LastClickedAppointment != null)
+        {
+            switch (clickedItem.CommandName.ToLowerInvariant())
+            {
+                case "delete":
+                    await DeleteAppt(LastClickedAppointment);
+                    break;
+                case "toggleimportant":
+                    await ToggleAppointmentImportant(LastClickedAppointment);
+                    break;
+                default:
+                    break;
+            }
+        }
+        LastClickedAppointment = null;
+    }
+
+    private async Task DeleteAppt(SchedulerAppointment appt)
+    {
+        if (appt.IsImportant)
+        {
+            return;
+        }
+        Appointments.Remove(appt);
+        await Task.Delay(100);
+        SchedulerRef.Rebind();
+    }
+
+    private async Task ToggleAppointmentImportant(SchedulerAppointment appt)
+    {
+        appt.IsImportant = !appt.IsImportant;
+        var index = Appointments.FindIndex(i => i.Id == appt.Id);
+        if (index != -1)
+        {
+            Appointments[index] = appt;
+        }
+        await Task.Delay(100);
```

Comment:
It's good to insert a comment why is the delay necessary and if the developer may need to increase it.